syntax = "proto3";
package modyn.storage;

service Storage {
  rpc Get (GetRequest) returns (stream GetResponse) {}
  rpc GetNewDataSince (GetNewDataSinceRequest) returns (stream GetNewDataSinceResponse) {}
  rpc CheckAvailability (DatasetAvailableRequest) returns (DatasetAvailableResponse) {}
  rpc RegisterNewDataset (RegisterNewDatasetRequest) returns (RegisterNewDatasetResponse) {}
}

message GetRequest {
  string dataset_id = 1;
  repeated string keys = 2;
}

message GetResponse {
  bytes chunk = 1;
}

message GetNewDataSinceRequest {
  string dataset_id = 1;
  int64 timestamp = 2;
}

message GetNewDataSinceResponse {
  repeated string keys = 1;
}

message DatasetAvailableRequest {
  string dataset_id = 1;
}

message DatasetAvailableResponse {
  bool available = 1;
}

message RegisterNewDatasetRequest {
  string dataset_id = 1;
  int32 filesystem_wrapper_type = 2;
  int32 file_wrapper_type = 3;
  string description = 4;
  string base_path = 5;
}

message RegisterNewDatasetResponse {
  bool success = 1;
}